tests/optim/test_layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,220 @@
+        self.relu2 = nn.ReLU()
+        self.fc3 = nn.Linear(self.hidden_size, 1)
+        self.sigmoid = nn.Sigmoid()
+        self.identity = nn.Identity()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:  # type: ignore
+        out = self.fc1(x)
+        out = self.relu1(out)
+        out = self.fc2(out)
+        out = self.relu2(out)
+        out = self.fc3(out)
+        out = self.sigmoid(out)
+        out = self.identity(out)
+        return out
+
+
+# assign labels
+def blob_label(y: np.ndarray, label: int, loc: List) -> np.ndarray:
+    target = np.copy(y)  # type: ignore
+    for l in loc:
+        target[y == l] = label
+    return target
+
+
+def load_data(model_type: str) -> Union[DataLoader, Tuple[Any, Any]]:
+    data = None
+    if model_type == "linear_model":
+        torch.manual_seed(11)
+        x_train, y_train = make_blobs(n_samples=40, n_features=2, cluster_std=1.5, shuffle=True, random_state=10)
+        x_train = torch.FloatTensor(x_train)
+        y_train = torch.FloatTensor(blob_label(y_train, 0, [0]))
+        y_train = torch.FloatTensor(blob_label(y_train, 1, [1, 2, 3]))
+
+        data = (x_train, y_train)
+
+    if model_type == "vision_model":
+        torch.manual_seed(10)
+        transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+        train_ds = torchvision.datasets.CIFAR10(root="./data", train=True, download=True, transform=transform)
+        train_ds_loader = torch.utils.data.DataLoader(train_ds, batch_size=128, shuffle=False, num_workers=2)
+
+        image, _ = train_ds[0]
+        assert image.shape == torch.Size([3, 32, 32])
+        data = train_ds_loader  # type: ignore
+    return data
+
+
+def train_linear_model(model: FeedForward, per_layer_scaling=False) -> FeedForward:
+    criterion = torch.nn.BCEWithLogitsLoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
+
+    x_train, y_train = load_data("linear_model")
+    num_epochs = 2
+    model.train()
+
+    layers_to_scale = {"fc1": 1024, "fc2": 512, "fc3": 1024} if per_layer_scaling else {}
+    layer_scaler = LayerwiseGradientScaler(model, layers_to_scale)
+
+    for _ in range(num_epochs):
+        optimizer.zero_grad()
+
+        # scale the gradients
+        layer_scaler.scale()
+
+        with autocast():
+            # forward pass
+            y_pred = model(x_train)
+            # compute loss
+            loss = criterion(y_pred.squeeze(), y_train)
+
+        loss.backward()
+
+        # unscale the gradients
+        layer_scaler.unscale()
+
+        # update weights and scaling factor
+        layer_scaler.step(optimizer)
+
+    return model
+
+
+def test_linear_model() -> None:
+    model1 = FeedForward(2, 10)
+    model2 = FeedForward(2, 10)
+
+    vanilla_model = train_linear_model(model1, False)
+    scaled_model = train_linear_model(model2, True)
+
+    def get_params_with_grad(trained_model):
```

Comment:
this can be factored out as a file level util i think.

Reply:
yes, nice catch. done.